Using virtual type to configure plugin, interceptor method cannot be …

GrassH opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes #33980 

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. rm generated/* -Rf && bin/magento test:plugin:calculate
Result is `0`;
2. rm generated/* -Rf && bin/magento setup:di:compile && bin/magento test:plugin:calculate
Result is `0`;

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

GrassH committed:
Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented on setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php:
```diff
@@ -201,6 +210,11 @@ private function getInterceptedMethods($interceptionConfiguration)
         foreach ($interceptionConfiguration as &$plugins) {
             $pluginsMethods = [];
             foreach ($plugins as $plugin) {
+                $realPlugin = $this->omConfig->getOriginalInstanceType($plugin);
+                if ($realPlugin != $plugin) {
+                    $plugin = $realPlugin;
+                }
```
Looks like it could be simplified
```suggestion
                $plugin = $this->omConfig->getOriginalInstanceType($plugin);
```

GrassH committed:
Update setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php

Co-authored-by: Ihor Sviziev <ihor-sviziev@users.noreply.github.com>

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @GrassH,
Could you check failing unit tests and cover your changes with some test?

GrassH committed:
Added tests for #33981 (https://github.com/magento/magento2/issues/33980)

GrassH committed:
Merge branch '2.4-develop' into develop

GrassH commented:
Hi @ihor-sviziev, added tests, please check.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Using virtual type to configure plugin, interceptor method cannot be generated correctly in setup:di:compile command

1. Make changes in constructor backward compatible
2. Fix static tests
- extract to separate method
- replace array_merge in loop with a single array_merge

Den4ik reviewed (changes requested):
@GrassH Thanks for PR.
Could you please fix failed Static tests.
Other failed tests are not related to PR and should be grean after some re-runs

ihor-sviziev commented:
@magento run all tests

Den4ik reviewed (approved):
@GrassH Changes looks ok for me. Failed tests related to problems on Magento CI side.
@ihor-sviziev 👍 thanks for helps with this PR

magento-engcom-team commented:
Hi @Den4ik, thank you for the review.
[ENGCOM-9217](https://jira.corp.magento.com/browse/ENGCOM-9217) has been created to process this Pull Request

magento-engcom-team commented:
@GrassH thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

Den4ik commented on setup/src/Magento/Setup/Module/Di/Code/Generator/InterceptionConfigurationBuilder.php:
```diff
@@ -10,9 +10,14 @@
+use Magento\Framework\Interception\ObjectManager\ConfigInterface;
 use Magento\Framework\ObjectManager\InterceptableValidator;
+use Magento\Setup\Module\Di\Code\Reader\Type;
+use function array_merge;
+use function array_unique;
```
I'm not sure that it's really needed here

Den4ik replied:
```suggestion
```

ihor-sviziev replied:
Fixed

hostep replied:
These probably were added for increased performance.
I'm going to link just [one blog](https://www.bram.us/2020/04/18/php-performance-tip-use-fully-qualified-function-calls-when-working-with-namespaces/), but there is plenty of information about this trick available on the web.

Not sure how much it will matter in this particular case though ...

Den4ik replied:
@hostep I don't see any benchmark results for language constructions.
It's interested and I'll check it when I have time

hostep replied:
The easiest way to test it would probably be to install a Magento version using composer and also install the following composer plugin: https://github.com/dg/composer-backslasher and then check performance differences somehow.
I have no experience with this plugin and I don't know if it will work properly on all Magento modules and their dependencies. But it's probably worth taking a shot.

Anyways, this is completely off topic in this PR 😆

Den4ik reviewed (changes requested):
Sorry, just rechecked changes and see that I overlooked some additionals. Please look at my comment

ihor-sviziev commented:
@magento run all tests

sivaschenko commented:
@ihor-sviziev @GrassH the cause of failing functional tests appears to be an issue with running `bin/mftf generate:tests`

GrassH commented:
@magento run Functional Tests CE

GrassH commented:
@magento give me test instance

GrassH commented:
@magento give me test instance

GrassH commented:
@magento give me test instance

GrassH commented:
@magento give me test instance

GrassH commented:
@magento give me test instance

Den4ik commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE, Sample Data Tests CE, Sample Data Tests B2B, Sample Data Tests EE

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE, Sample Data Tests CE, Sample Data Tests B2B, Sample Data Tests EE

ihor-sviziev reviewed (changes requested):
Please resolve conflicts and failing tests

ishakhsuvarov commented:
Hi

I am closing this PR for now as it seems like it has become stale and still needs work.

Please feel free to reopen if you wish to continue.

Thanks!
